File: 2704-maximum-difference-by-remapping-a-digit/maximum-difference-by-remapping-a-digit.rb
# @param {Integer} num
# @return {Integer}
def min_max_difference(num)
    first_non_nine = nil
    first_digit = nil
    remaining = num

    while remaining > 0
        digit = remaining % 10
        first_non_nine = digit if digit != 9
        first_digit = digit
        remaining /= 10
    end

    remaining = num
    min_val = 0
    max_val = 0
    multiplier = 1

    # Tạo số min và max
    while remaining > 0
        digit = remaining % 10
        min_digit = digit
        max_digit = digit

        min_digit = 0 if digit == first_digit
        max_digit = 9 if digit == first_non_nine

        min_val += multiplier * min_digit
        max_val += multiplier * max_digit

        multiplier *= 10
        remaining /= 10
    end

    max_val - min_val
end

File: 2704-maximum-difference-by-remapping-a-digit/maximum-difference-by-remapping-a-digit.go
func minMaxDifference(num int) int {
	firstNonNine, firstDigit := -1, -1
	remaining := num

	// Tìm chữ số đầu tiên khác 9 và chữ số đầu tiên (từ phải sang)
	for remaining > 0 {
		digit := remaining % 10
		if digit != 9 {
			firstNonNine = digit
		}
		firstDigit = digit
		remaining /= 10
	}

	remaining = num
	minVal, maxVal := 0, 0
	multiplier := 1

	// Tạo số min và max
	for remaining > 0 {
		digit := remaining % 10
		minDigit, maxDigit := digit, digit

		if digit == firstDigit {
			minDigit = 0
		}
		if digit == firstNonNine {
			maxDigit = 9
		}

		minVal += multiplier * minDigit
		maxVal += multiplier * maxDigit

		multiplier *= 10
		remaining /= 10
	}

	return maxVal - minVal
}